snap/info_snap_yaml.go

```diff
@@ -21,18 +21,200 @@ package snap
+	for name, app := range y.RawApps {
+		snap.Apps[name] = &AppInfo{
+			Snap:  snap,
+			Name:  name,
+			slots: app.SlotNames,
+			plugs: app.PlugNames,
+		}
+	}
+	// Collect app-level implicit definitions of plugs and slots
+	for _, app := range y.RawApps {
+		for _, name := range app.PlugNames {
+			if _, ok := snap.Plugs[name]; !ok {
+				snap.Plugs[name] = &PlugInfo{
+					Snap:      snap,
+					Name:      name,
+					Interface: name,
+				}
+			}
+		}
+		for _, name := range app.SlotNames {
+			if _, ok := snap.Slots[name]; !ok {
+				snap.Slots[name] = &SlotInfo{
+					Snap:      snap,
+					Name:      name,
+					Interface: name,
+				}
+			}
+		}
+	}
+	// Bind apps to plugs and slots
+	for appName, app := range y.RawApps {
+		for _, slotName := range app.SlotNames {
+			slot := snap.Slots[slotName]
+			slot.apps = append(slot.apps, appName)
+		}
+		for _, plugName := range app.PlugNames {
+			plug := snap.Plugs[plugName]
+			plug.apps = append(plug.apps, appName)
+		}
+	}
+	// Bind unbound plugs and slots to all apps
+	for _, plug := range snap.Plugs {
+		if len(plug.apps) == 0 {
+			for name := range y.RawApps {
+				plug.apps = append(plug.apps, name)
+			}
+		}
+	}
+	for _, slot := range snap.Slots {
+		if len(slot.apps) == 0 {
+			for name := range y.RawApps {
+				slot.apps = append(slot.apps, name)
+			}
+		}
+	}
+	// Bind unbound apps to all plugs and slots
+	for _, app := range snap.Apps {
+		// NOTE: This used RawPlugs and RawSlots so that implicitly defined
+		// (non-top-level) plugs and slots don't get bound here.
+		if len(app.plugs) == 0 {
+			for name := range y.RawPlugs {
+				app.plugs = append(app.plugs, name)
+			}
+		}
+		if len(app.slots) == 0 {
+			for name := range y.RawSlots {
+				app.slots = append(app.slots, name)
+			}
+		}
+	}
 	// FIXME: validation of the fields
+	return snap, nil
+}
 
-	return &s, nil
+func convertToSlotOrPlugData(plugOrSlot, name string, data interface{}) (string, map[string]interface{}, error) {
+	switch data.(type) {
+	case map[interface{}]interface{}:
+		attrs := make(map[string]interface{})
+		iface := ""
+		for keyData, valueData := range data.(map[interface{}]interface{}) {
+			key, ok := keyData.(string)
+			if !ok {
+				return "", nil, fmt.Errorf("%s %q has attribute that is not a string (found %T)",
+					plugOrSlot, name, keyData)
+			}
+			if strings.HasPrefix(key, "$") {
+				return "", nil, fmt.Errorf("%s %q uses reserved attribute %q", plugOrSlot, name, key)
+			}
+			// XXX: perhaps we could special-case "label" the same way?
+			if key == "interface" {
+				value, ok := valueData.(string)
+				if !ok {
+					return "", nil, fmt.Errorf("interface name on %s %q is not a string (found %T)",
+						plugOrSlot, name, valueData)
+				}
+				iface = value
+			} else {
+				attrs[key] = valueData
+			}
+		}
+		if len(attrs) == 0 {
+			attrs = nil
+		}
+		if iface == "" {
+			return "", nil, fmt.Errorf("%s %q doesn't define interface name", plugOrSlot, name)
```

Comment:
Yes but according the the syntax in the google doc, if you define a structure you have to have the field set explicitly. If I read that right it is easy to correct.